server/coordinator/procedure/transfer_leader.go

```diff
@@ -0,0 +1,317 @@
+		return
+	}
+}
+
+func transferLeaderCloseOldLeaderCallback(event *fsm.Event) {
+	request := event.Args[0].(*TransferLeaderCallbackRequest)
+	ctx := request.ctx
+
+	closeShardRequest := eventdispatch.CloseShardRequest{
+		ShardID: uint32(request.shardID),
+	}
+	if err := request.dispatch.CloseShard(ctx, request.oldLeaderNodeName, closeShardRequest); err != nil {
+		cancelEventWithLog(event, err, "close shard failed", zap.Uint32("shardId", uint32(request.shardID)), zap.String("oldLeaderName", request.oldLeaderNodeName))
+		return
+	}
+}
+
+func transferLeaderOpenNewShardCallback(event *fsm.Event) {
+	request := event.Args[0].(*TransferLeaderCallbackRequest)
+	ctx := request.ctx
+
+	// TODO: Shard version is only exists in ShardTables now, it is not sensible. Because shard version will be update when shard node mapping changed or shard table mapping changed, so ShardNodes also need shard version, consider refactor later.
+	shardTablesMapping := request.cluster.GetTables([]storage.ShardID{request.shardID}, request.oldLeaderNodeName)
+	shardTables, exists := shardTablesMapping[request.shardID]
+	if !exists {
+		cancelEventWithLog(event, cluster.ErrTableNotFound, "tables on shard not found")
+		return
+	}
+
+	openShardRequest := eventdispatch.OpenShardRequest{
+		Shard: cluster.ShardInfo{ID: request.shardID, Role: storage.ShardRoleLeader, Version: shardTables.Shard.Version + 1},
+	}
+
+	if err := request.dispatch.OpenShard(ctx, request.newLeaderNodeName, openShardRequest); err != nil {
+		cancelEventWithLog(event, err, "open shard failed", zap.Uint32("shardId", uint32(request.shardID)), zap.String("newLeaderNode", request.newLeaderNodeName))
+		return
+	}
+}
+
+func transferLeaderFinishCallback(event *fsm.Event) {
+	request := event.Args[0].(*TransferLeaderCallbackRequest)
+	log.Info("transfer leader finish", zap.Uint32("shardID", uint32(request.shardID)), zap.String("oldLeaderNode", request.oldLeaderNodeName), zap.String("newLeaderNode", request.newLeaderNodeName))
+}
+
+func (p *TransferLeaderProcedure) updateStateWithLock(state State) {
+	p.lock.Lock()
+	p.state = state
+	p.lock.Unlock()
+}
+
+func (p *TransferLeaderProcedure) convertToMeta() (Meta, error) {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	rawData := TransferLeaderProcedurePersistRawData{
+		ID:                p.id,
+		FsmState:          p.fsm.Current(),
+		ShardID:           p.shardID,
+		OldLeaderNodeName: p.oldLeaderNodeName,
+		NewLeaderNodeName: p.newLeaderNodeName,
+		State:             p.state,
+	}
+	rawDataBytes, err := json.Marshal(rawData)
+	if err != nil {
+		return Meta{}, errors.WithMessage(err, "marshal raw data failed")
+	}
+
+	meta := Meta{
+		ID:      p.id,
+		Typ:     TransferLeader,
+		State:   p.state,
+		RawData: rawDataBytes,
+	}
+
+	return meta, nil
+}
+
+func ConvertToTransferLeaderProcedure(meta Meta, storage Storage, dispatch eventdispatch.Dispatch, cluster *cluster.Cluster) (*TransferLeaderProcedure, error) {
```

Comment:
The function is useless?

Reply:
It will be used by scheduler when read procedure metadata from storage and  restart procedure.